File: frontend/lib/api/projects.ts
import type { Project, ProjectCreateInput, ProjectUpdateInput } from "@/types";
import { apiFetch } from "./client";

export function listProjects(signal?: AbortSignal): Promise<Project[]> {
  return apiFetch<Project[]>("/projects", { signal });
}

export function getProject(projectId: number, signal?: AbortSignal): Promise<Project> {
  return apiFetch<Project>(`/projects/${projectId}`, { signal });
}

export function createProject(input: ProjectCreateInput): Promise<Project> {
  return apiFetch<Project>("/projects", { method: "POST", body: input });
}

export function updateProject(projectId: number, input: ProjectUpdateInput): Promise<Project> {
  return apiFetch<Project>(`/projects/${projectId}`, { method: "PATCH", body: input });
}

export function deleteProject(projectId: number): Promise<void> {
  return apiFetch<void>(`/projects/${projectId}`, { method: "DELETE" });
}

File: backend/app/api/routes/projects.py
"""Project (product brief) routes."""

from fastapi import APIRouter, Depends, status
from sqlalchemy.orm import Session

from app.database.session import get_db
from app.models.project import Project
from app.schemas.project import ProjectCreate, ProjectRead, ProjectUpdate
from app.services.project import ProjectService

router = APIRouter(prefix="/projects", tags=["projects"])


def get_project_service(db: Session = Depends(get_db)) -> ProjectService:
    return ProjectService(db)


@router.post("", response_model=ProjectRead, status_code=status.HTTP_201_CREATED)
def create_project(
    payload: ProjectCreate, service: ProjectService = Depends(get_project_service)
) -> Project:
    return service.create(payload)


@router.get("", response_model=list[ProjectRead])
def list_projects(service: ProjectService = Depends(get_project_service)) -> list[Project]:
    return service.list()


@router.get("/{project_id}", response_model=ProjectRead)
def get_project(project_id: int, service: ProjectService = Depends(get_project_service)) -> Project:
    return service.get(project_id)


@router.patch("/{project_id}", response_model=ProjectRead)
def update_project(
    project_id: int,
    payload: ProjectUpdate,
    service: ProjectService = Depends(get_project_service),
) -> Project:
    return service.update(project_id, payload)


@router.delete("/{project_id}", status_code=status.HTTP_204_NO_CONTENT)
def delete_project(project_id: int, service: ProjectService = Depends(get_project_service)) -> None:
    service.delete(project_id)
